### PR TITLE
librbd: Modify locks and atomics in AioCompletion to improve performance

### PR DESCRIPTION
**Pull request by adam-lj**
This PR proposes the following performance improvements to AioCompletion:

1. Replace the conditional variable with the atomic wait/notify functions introduced in C++20.
2. Remove the mutex lock which was previously required to update the conditional variable.
3. Relax the memory order of the atomic operations to update the AioCompletion state.

Flame graph analysis of `rbd bench` indicates an almost 1% performance improvement on average in the function AioCompletion::notify_callbacks_complete().

Before:
AioCompletion_before

After:
AioCompletion_after

This data was captured by running a workload of 4K sequential writes to a 200G volume.

I have ran `bin/ceph_test_librbd `to validate that the code reliant on the conditional variable still works and specifically ran -`-gtest_filter=TestLibRBD.TestIO` with valgrind to validate that there are no memory leaks.

This is my first pull request for ceph so I'm open to feedback, please let me know if you think there's anything I've missed.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by adam-lj** [comment]
jenkins test make check arm64

**Comment by adam-lj** [comment]
The arm64 make check failed on flap.sh which is known to be broken: https://tracker.ceph.com/issues/66556

All the other tests appear to have passed.

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [inline src/librbd/io/AioCompletion.cc:280]: Is accessing `cond` without holding the mutex safe?  Can the assignment to `cond` in `wait_for_complete()` be reordered with the initial test on `state`/become visible to the thread executing this condition late, so that the thread executing `wait_for_complete()` ends up waiting on `cond` but we never notify it because `cond` happened to test NULL?

Comment (idryomov) [reply src/librbd/io/AioCompletion.cc:280]: > Can the assignment to `cond` in `wait_for_complete()` be reordered with the initial test on `state`/become visible to the thread executing this condition late

If not, wouldn't it be reported as a data race anyway?

File: src/librbd/io/AioCompletion.h
Comment (idryomov) [inline src/librbd/io/AioCompletion.h:48]: > Modify the conditional variable to be a unique_ptr so that we only allocate memory when it is actually needed (rarely since it only appears to be used in test code).

`AioCompletion::wait_for_complete()` implements the public API (`rbd_aio_wait_for_complete()` and the corresponding C++ variant), so we don't really control its usage.  In-tree, apart from test code, it's used in the implementation of `rbd ls` command (see `src/tools/rbd/action/List.cc`).  In general, I'd agree with the "rarely" estimate.

Instead of involving a dynamic allocation, I wonder if we can get rid of `std::condition_variable` and the associated locking altogether by utilizing `wait()` and `notify_all()` methods directly on `state` (new in C++20):

https://en.cppreference.com/w/cpp/atomic/atomic/wait
https://en.cppreference.com/w/cpp/atomic/atomic/notify_all

File: src/librbd/io/AioCompletion.cc
Comment (adam-lj) [reply src/librbd/io/AioCompletion.cc:280]: Good point about the race hazard. Would you recommend moving the nullptr check to after the mutex, or removing it entirely?

I don't think it's possible for cond to be null if waiting_for_complete is true, but I thought it would be good practice to add a nullptr check anyway.

File: src/librbd/io/AioCompletion.h
Comment (adam-lj) [reply src/librbd/io/AioCompletion.h:48]: Sorry, that was poorly worded as I knew about the usage in `rbd ls` as well! I wasn't aware of the public API usage though so thank you for bringing that to my attention.

Replacing the conditional variable with the new C++20 atomic wait/notify functions is a great idea, I will try that and take some more measurements to see if it gives us any further performance improvements.

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [reply src/librbd/io/AioCompletion.cc:280]: > Would you recommend moving the nullptr check to after the mutex, or removing it entirely?

With the code as is, the nullptr check can't removed entirely because of another race.  Consider:
- `wait_for_complete()` sets `waiting_for_complete` to true but doesn't get a change to grab the mutex
- `notify_callbacks_complete()` loads `waiting_for_complete`, sees that it's true, grabs the mutex and attempts to call `notify_all()` on `cond` that hasn't even been allocated, let alone "published"

I'm hoping we can get rid of all of this with the switch to C++20 atomic wait/notify.

Comment (idryomov) [reply src/librbd/io/AioCompletion.cc:280]: > I don't think it's possible for cond to be null if waiting_for_complete is true

It seems possible to me, as a result of the race that I outlined above.

Comment (adam-lj) [reply src/librbd/io/AioCompletion.cc:280]: Yes, sorry - I meant not possible under normal circumstances, but obviously it is possible in the race condition.

Good point that we won't need the mutex with the C++20 atomic wait/notify so I will make that change first.

**Comment by adam-lj** [comment]
I've just implemented and tested the requested changes to use the new C++20 atomic functions, the performance gain seems similar for both.

**Comment by Svelar** [comment]
jenkins test make check arm64

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [inline src/librbd/io/AioCompletion.cc:38]: > Looks like there is a unit test failure sorry, debugging...

The signature is
```
void wait( T old, std::memory_order order =
                      std::memory_order::seq_cst ) const noexcept;
```
and it blocks if `old` matches the current value.  So it looks like you end up blocking when `state` is already `AIO_STATE_COMPLETE` instead of the other way around.

Comment (adam-lj) [reply src/librbd/io/AioCompletion.cc:38]: Yes I've just realised that and I'm testing the fix now.

Comment (adam-lj) [reply src/librbd/io/AioCompletion.cc:38]: So that was a silly mistake on my part, sorry - I misunderstood the atomic wait function as "wait until the state is equal to this value", rather than "wait until the state is NOT equal to this value" (I am a C++ newbie if you can't tell!).

I've just changed it to AIO_STATE_PENDING and ran all the RBD unit tests locally with no errors, so that seems to have fixed it. Is there an edge case though where the state is AIO_STATE_CALLBACK, or does that not matter since I only call notify_all after it gets set to AIO_STATE_COMPLETE?

Comment (idryomov) [reply src/librbd/io/AioCompletion.cc]: > Is there an edge case though where the state is AIO_STATE_CALLBACK, or does that not matter since I only call notify_all after it gets set to AIO_STATE_COMPLETE?

This is a good question!  I was going to answer "yes", because a user could call `AioCompletion::wait_for_complete()` while the state is `AIO_STATE_CALLBACK` (i.e. while callbacks are being run) and we don't want to return prematurely in that case.  There was an explicit fix to that effect, see commit 77f7c1cf829e5417eb356ebf3bf1fbc8ff7a25a5.

However, looking at today's code, the distinction between `AIO_STATE_CALLBACK` and `AIO_STATE_COMPLETE` doesn't seem to be material anymore:
- The entire point of  commit 77f7c1cf829e5417eb356ebf3bf1fbc8ff7a25a5 was this bit in `AioCompletion::start_op()`, but it was rewritten in commit 002afa0fe37544ab4eb15f973e00068e1240c0ef:

  ```
      Mutex::Locker locker(lock);
      if (state == STATE_PENDING && !async_op.started()) {
        async_op.start_op(*ictx);
      }
  ```

- The same commit 002afa0fe37544ab4eb15f973e00068e1240c0ef changed `AioCompletion::is_complete()` to test for `!= AIO_STATE_PENDING` instead of `== AIO_STATE_COMPLETE`.  No way to tell whether it was intended or a typo, but I think that it's a mistake that needs to be fixed because it made `AioCompletion::wait_for_complete()` and `AioCompletion::is_complete()` inconsistent with each other -- their meaning of "complete" is different in today's code.

With the above in mind, it seems like we can go ahead and revert state to a bool, fixing that mistake and simplifying at the same time:

- `done` would be set to true where state is set to `AIO_STATE_COMPLETE` today
-  you would be able to call `done.wait(false, ...)` without worrying about edge cases here
- `AioCompletion::is_complete()` would return `done`

What do you think?

Comment (adam-lj) [reply src/librbd/io/AioCompletion.cc]: Yes that sounds good to me, I just searched the code for AIO_STATE_CALLBACK and can't find anywhere it is actually used anymore so was about to suggest the same.

**Comment by adam-lj** [comment]
jenkins test make check

**Comment by adam-lj** [comment]
jenkins test api

**Comment by adam-lj** [comment]
jenkins test make check

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [inline src/librbd/io/AioCompletion.cc:98]: This should be done in `notify_callbacks_complete()` because a user could call `wait_for_complete()` while the callbacks are being run and we don't want to return prematurely in that case:
> There was an explicit fix to that effect, see commit [77f7c1c](https://github.com/ceph/ceph/commit/77f7c1cf829e5417eb356ebf3bf1fbc8ff7a25a5).

> * `done` would be set to true where state is set to `AIO_STATE_COMPLETE` today

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [inline src/librbd/io/AioCompletion.cc:275]: Do we still need to bother with `waiting_for_complete` now that that there is no mutex around here?  Can you see if notifying `cond` unconditionally makes a difference in your measurements?

It would be nice to update the description of the PR and post a new set of flamegraphs since the approach has changed quite radically.  Of the three changes that are being talked about there, only one is up-to-date:
> 3. Relax the memory order of the atomic operations to update the AioCompletion state.

File: src/librbd/io/AioCompletion.h
Comment (idryomov) [inline src/librbd/io/AioCompletion.h:55]: The comment above is no longer true -- we have grown wait/notify use.  Let's remove it.

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [inline src/librbd/io/AioCompletion.cc:39]: These braces were defining the critical section of the mutex.  They are useless now -- let's remove them and reindent.

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [inline src/librbd/io/AioCompletion.cc:242]: It seems like `complete` would be a better name for the class member as well.  I used `done` in my comment only because it's how it was named before commit 77f7c1cf829e5417eb356ebf3bf1fbc8ff7a25a5:
```suggestion
  bool complete = complete.load(std::memory_order_acquire);
```

**Comment by adam-lj** [comment]
jenkins test make check

File: src/librbd/io/AioCompletion.cc
Comment (adam-lj) [reply src/librbd/io/AioCompletion.cc:98]: That was what I originally did, but then I noticed that some unit tests were failing because the complete state was still set to false when they called `is_complete()`, which led me to believe that the `!= AIO_STATE_PENDING` check was actually important to those tests.

I supposes I could update the state in both `complete()` and `notify_callbacks_complete()` to cover all the bases, if you don't think that would be redundant?

File: src/librbd/io/AioCompletion.cc
Comment (adam-lj) [reply src/librbd/io/AioCompletion.cc:275]: The original reason for `waiting_for_complete` was to avoid the performance cost of taking the lock and notifying the conditional in the normal I/O path (e.g. when running `rbd bench`, as I did to take these measurements) where the conditional is not actually used.

Removing the lock partially alleviates that problem, but I'm guessing there is still a performance cost to using the atomic wait+notify when it's not needed? I can try it though and take some more measurements to confirm.

I did take some more measurements and flame graphs with the new approach and found that the performance gain was more or less the same for both, since either way `waiting_for_complete` prevents us from going down the conditional path when running `rbd bench` (so in the original approach we don't take the lock, in the new approach we don't do the atomic wait+notify).

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [reply src/librbd/io/AioCompletion.cc:98]: > That was what I originally did, but then I noticed that some unit tests were failing because the complete state was still set to false when they called `is_complete()`, which led me to believe that the `!= AIO_STATE_PENDING` check was actually important to those tests

Do you have a list of those tests?
 
> I supposes I could update the state in both `complete()` and `notify_callbacks_complete()` to cover all the bases, if you don't think that would be redundant?

That would effectively throw out the fix made in commit 77f7c1cf829e5417eb356ebf3bf1fbc8ff7a25a5.

Comment (adam-lj) [reply src/librbd/io/AioCompletion.cc:98]: I reran the make check a couple of times and it seems to fail on a different one each time, so it's not consistent and I guess there must be some sort of race condition there.

I didn't make a note of all of them sorry, but I do know that `TestLibRBD.ImagePollIO` was one of them (because I reran it locally and was unable to recreate).

Comment (idryomov) [reply src/librbd/io/AioCompletion.cc:98]: I think I see the issue -- the event socket is notified before `done`/`complete` is set.  I have opened https://github.com/ceph/ceph/pull/58591 to clean that up and separate dealing with inconsistencies in the existing code from the conversion to atomic wait/notify to gain some performance.  Let's see if tests pass there!

**Comment by adam-lj** [comment]
I've made the requested changes and updated the description, so once https://github.com/ceph/ceph/pull/58591 I will merge your changes in @idryomov and retest. This includes a new set of flame graphs which seems to indicate a similar performance improvement with and without the `waiting_for_complete` boolean, so I agree that this is no longer required.

Please let me know if you have any further feedback.

**Comment by adam-lj** [comment]
jenkins test api

**Comment by adam-lj** [comment]
jenkins test make check arm64

**Comment by adam-lj** [comment]
jenkins test make check

**Comment by adam-lj** [comment]
jenkins test make check arm64

**Comment by adam-lj** [comment]
jenkins test make check arm64

**Comment by adam-lj** [comment]
@idryomov Are you happy to approve this now please?

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [inline src/librbd/io/AioCompletion.cc:236]: Let's match the exit tracepoint and the return value, as it was before:
```suggestion
  tracepoint(librbd, aio_is_complete_exit, done);
```

File: src/librbd/io/AioCompletion.cc
Comment (idryomov) [inline src/librbd/io/AioCompletion.cc:261]: I'd place this call where the corresponding call on `cond` was.  The event socket has always been notified first, so unless there is a good rationale for the change it's better to keep it that way:
```
    ictx->event_socket.notify();
  }

  completed.notify_all();

  if (image_dispatcher_ctx != nullptr) {
```

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
@adam-lj This hasn't been through the test suite yet -- one of the testing labels just got appended.  In the future, please leave merging PRs to respective maintainers.

**Comment by adam-lj** [comment]
> @adam-lj This hasn't been through the test suite yet -- one of the testing labels just got appended. In the future, please leave merging PRs to respective maintainers.

Sorry I assumed since you had approved it was ready to merge! I will hold off in future.

**Pull request merged**
